lib/canister.rb

```diff
@@ -53,9 +56,13 @@ def register(key, &block)
   # @param key [Symbol]
   def resolve(key)
     add_dependent(key)
-    stack << key
+    @mutex.synchronize do
+      stack << key
+    end
     value = resolved[key.to_sym] ||= registry[key.to_sym].call(self)
```

Comment:
This assigns to `resolved`, so we probably need to synchronize there too. In general, I'm somewhat concerned that with all the locking & unlocking here it increases the opportunities for mysterious deadlocks or for other threads to interrupt in the middle of resolving something & break things. If there was an atomic way to say "get the lock if we don't already have it", that would be preferable. However, it looks like if we already hold the mutex and do `@mutex.synchronize` it will complain... I wonder if it would work if we do something like:

```ruby
def synchronize
  if @mutex.owned
    yield
  else
    @mutex.synchronize { yield }
  end
end

def register(key, &block)
  synchronize do 
    invalidate(key) if registered?(key)
    registry[key.to_sym] = block
  end
  self
end 

def resolve(key)
  synchronize do
    stack << key
    value = resolved[key.to_sym] ||= registry[key.to_sym].call(self)
    stack.pop
  end
  value
end
```

I'm not sure what the

Reply:
Using `synchronize` looks good to me. We're not handling an infinite loop where the recursive calls to `resolve` actually contain a truly recursive call (e.g., `Services.register(:recur) { Services[:other} + Services[:recur]  }`) or mutually recursive definitions (where `other` relies on `recur` and vice-versa) but ... don't do that.  Or we could do 

```ruby
def resolve(key)
    key = key.to_sym
    if stack.include? key
      loop_trace = stack.slice(stack.find_index(key)..-1) + [key]
      raise "Infinite loop in definition: value of #{key} refers to itself via #{loop_trace.join("->")}" 
    end
   ....
end
```

The other thing we should do is to remove the `attr_reader` line and refer to all the underlying data structures only by their instance variable (e.g., `@stack`, `@resolved`). With the `attr_reader` they're technically part of the public interface and we don't want anyway mucking around with them directly. There's no reason anyone would want to _do_ that,, but we might as well lock it down.

Reply:
Oh, it also feels like the use of `key.to_sym` isn't consistent? I wonder if every method should just start with `key = key.to_sym`.

Reply:
@billdueber `attr_reader` is inside the `private` section so at least on Ruby 3.1 I am not able to use those accessors from for example `irb`. That being said, I think all that is needed to enforce Symbol-only keys is adding that `#to_sym` at the head of `#register` and `#resolve` as the only public entrypoints that take a key value.
Patch forthcoming.

Reply:
Looks good to me at this point.